zcash_primitives/src/transaction/builder.rs

```diff
@@ -306,14 +440,76 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
             .ok_or(Error::InvalidAmount)
     }
 
+    pub fn compute_balances<BalanceError>(
+        &self,
+        fee_strategy: &impl FeeStrategy<Error = BalanceError>,
+    ) -> Result<TransactionBalances, BalanceError>
+    where
+        Error: From<BalanceError>,
+    {
+        fee_strategy.compute_balances(
+            self.transparent_builder.inputs(),
+            self.transparent_builder.outputs(),
+            self.sapling_builder.inputs(),
+            self.sapling_builder.outputs(),
+        )
+    }
+
+    #[cfg(feature = "zfuture")]
+    pub fn compute_balances_zfuture<BalanceError>(
+        &self,
+        fee_strategy: &impl FutureFeeStrategy<Error = BalanceError>,
+    ) -> Result<TransactionBalances, BalanceError>
+    where
+        Error: From<BalanceError>,
+    {
+        fee_strategy.compute_balances_zfuture(
+            self.transparent_builder.inputs(),
+            self.transparent_builder.outputs(),
+            self.sapling_builder.inputs(),
+            self.sapling_builder.outputs(),
+            self.tze_builder.inputs(),
+            self.tze_builder.outputs(),
+        )
+    }
+
     /// Builds a transaction from the configured spends and outputs.
     ///
     /// Upon success, returns a tuple containing the final transaction, and the
     /// [`SaplingMetadata`] generated during the build process.
-    pub fn build(
-        mut self,
+    pub fn build<BalanceError>(
+        self,
         prover: &impl TxProver,
-    ) -> Result<(Transaction, SaplingMetadata), Error> {
+        balance_calculator: &impl FeeStrategy<Error = BalanceError>,
+    ) -> Result<(Transaction, SaplingMetadata), Error>
+    where
+        Error: From<BalanceError>,
+    {
+        let balances = self.compute_balances(balance_calculator)?;
+        self.build_internal(prover, &balances)
+    }
+
+    #[cfg(feature = "zfuture")]
+    pub fn build_zfuture<BalanceError>(
+        self,
+        prover: &impl TxProver,
+        balance_calculator: &impl FutureFeeStrategy<Error = BalanceError>,
+    ) -> Result<(Transaction, SaplingMetadata), Error>
+    where
+        Error: From<BalanceError>,
+    {
+        let balances = self.compute_balances_zfuture(balance_calculator)?;
+        self.build_internal(prover, &balances)
+    }
+
+    fn build_internal<BalanceError>(
+        self,
+        prover: &impl TxProver,
+        balances: &TransactionBalances,
+    ) -> Result<(Transaction, SaplingMetadata), Error>
+    where
+        Error: From<BalanceError>,
+    {
```

Comment:
I don't understand why this generic is present at all; it isn't used anywhere AFAICT.
```suggestion
    fn build_internal(
        self,
        prover: &impl TxProver,
        balances: &TransactionBalances,
    ) -> Result<(Transaction, SaplingMetadata), Error> {
```